### PR TITLE
refactor(modules/datasource): fix null-check issues

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Strict null check implementation

## Context
- #7154
- requires #16150

### CODE REVIEW

File: lib/modules/datasource/cdnjs/index.spec.ts [hunk 1]
```diff
@@ -16,7 +16,10 @@ describe('modules/datasource/cdnjs/index', () => {
     });
 
     it('throws for empty result', async () => {
-      httpMock.scope(baseUrl).get(pathFor('foo/bar')).reply(200, null);
+      httpMock
+        .scope(baseUrl)
+        .get(pathFor('foo/bar'))
+        .reply(200, null as never);
```
Comment (viceice) [inline lib/modules/datasource/cdnjs/index.spec.ts:22]: null isn't allowed / valid
```suggestion
        .reply(200);
```

Use this to force got to throw an error
```suggestion
        .reply(200, '{');
```

File: lib/modules/datasource/clojure/index.spec.ts [hunk 2]
```diff
@@ -204,11 +204,11 @@ describe('modules/datasource/clojure/index', () => {
     httpMock
       .scope('https://failed_repo')
       .get('/org/example/package/maven-metadata.xml')
-      .reply(404, null);
+      .reply(404, null as never);
```
Comment (viceice) [inline lib/modules/datasource/clojure/index.spec.ts:207]: see other comment

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/dart/index.spec.ts [hunk 3]
```diff
@@ -10,7 +10,10 @@ const baseUrl = 'https://pub.dartlang.org/api/packages/';
 describe('modules/datasource/dart/index', () => {
   describe('getReleases', () => {
     it('returns null for empty result', async () => {
-      httpMock.scope(baseUrl).get('/non_sense').reply(200, null);
+      httpMock
+        .scope(baseUrl)
+        .get('/non_sense')
+        .reply(200, null as never);
```
Comment (viceice) [inline lib/modules/datasource/dart/index.spec.ts:16]:
```suggestion
      httpMock.scope(baseUrl).get('/non_sense').reply(200, '}');
```

File: lib/modules/datasource/datasource.spec.ts [hunk 4]
```diff
@@ -18,7 +18,7 @@ class TestDatasource extends Datasource {
     } catch (err) {
       this.handleGenericErrors(err);
     }
-    return Promise.resolve(undefined);
+    return Promise.resolve(undefined as never);
```
Comment (viceice) [inline lib/modules/datasource/datasource.spec.ts:21]: use `null`, is it's the allowed value
https://github.com/renovatebot/renovate/blob/9e28ef32752536501af3fbff8a63ccbe5343591c/lib/modules/datasource/types.ts#L78

File: lib/modules/datasource/datasource.spec.ts [hunk 5]
```diff
@@ -28,8 +28,8 @@ describe('modules/datasource/datasource', () => {
 
     httpMock.scope(exampleUrl).get('/').reply(429);
 
-    await expect(testDatasource.getReleases(undefined)).rejects.toThrow(
-      EXTERNAL_HOST_ERROR
-    );
+    await expect(
+      testDatasource.getReleases(undefined as never)
+    ).rejects.toThrow(EXTERNAL_HOST_ERROR);
```
Comment (viceice) [inline lib/modules/datasource/datasource.spec.ts:33]:
```suggestion
    await expect(
      testDatasource.getReleases(partial<GetReleasesConfig>({}))
    ).rejects.toThrow(EXTERNAL_HOST_ERROR);
```

File: lib/modules/datasource/galaxy-collection/index.spec.ts [hunk 6]
```diff
@@ -124,7 +124,7 @@ describe('modules/datasource/galaxy-collection/index', () => {
       expect(
         await getPkgReleases({
           datasource,
-          depName: null,
+          depName: null as never,
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.spec.ts:127]:
```suggestion
          depName: '',
```

File: lib/modules/datasource/git-refs/index.spec.ts [hunk 7]
```diff
@@ -17,7 +17,7 @@ describe('modules/datasource/git-refs/index', () => {
     it('returns nil if response is wrong', async () => {
       simpleGit.mockReturnValue({
         listRemote() {
-          return Promise.resolve(null) as Response<string>;
+          return Promise.resolve(null) as unknown as Response<string>;
```
Comment (viceice) [inline lib/modules/datasource/git-refs/index.spec.ts:20]:
```suggestion
          return Promise.resolve('') as Response<string>;
```

https://github.com/renovatebot/renovate/blob/de70ecad0409d99e3ae548deac7c6ef3e662b02d/lib/modules/datasource/git-refs/base.ts#L33-L38

File: lib/modules/datasource/git-tags/index.spec.ts [hunk 8]
```diff
@@ -18,7 +18,7 @@ describe('modules/datasource/git-tags/index', () => {
     it('returns nil if response is wrong', async () => {
       simpleGit.mockReturnValue({
         listRemote() {
-          return Promise.resolve(null) as Response<string>;
+          return Promise.resolve(null) as unknown as Response<string>;
```
Comment (viceice) [inline lib/modules/datasource/git-tags/index.spec.ts:21]:
```suggestion
          return Promise.resolve('') as Response<string>;
```

File: lib/modules/datasource/github-releases/common.spec.ts [hunk 9]
```diff
@@ -8,7 +8,7 @@ describe('modules/datasource/github-releases/common', () => {
     });
 
     it('defaults to github.com', () => {
-      const sourceUrl = getSourceUrlBase(null);
+      const sourceUrl = getSourceUrlBase(null as never);
```
Comment (viceice) [inline lib/modules/datasource/github-releases/common.spec.ts:11]:
```suggestion
      const sourceUrl = getSourceUrlBase(undefined);
```

File: lib/modules/datasource/github-tags/index.spec.ts [hunk 10]
```diff
@@ -40,7 +40,7 @@ describe('modules/datasource/github-tags/index', () => {
         .get(`/repos/${packageName}/commits?per_page=1`)
         .reply(200, [{ sha: 'abcdef' }]);
 
-      const res = await github.getDigest({ packageName }, null);
+      const res = await github.getDigest({ packageName }, null as never);
```
Comment (viceice) [inline lib/modules/datasource/github-tags/index.spec.ts:43]:
```suggestion
      const res = await github.getDigest({ packageName }, undefined);
```

File: lib/modules/datasource/github-tags/index.spec.ts [hunk 11]
```diff
@@ -51,7 +51,7 @@ describe('modules/datasource/github-tags/index', () => {
         .get(`/repos/${packageName}/commits?per_page=1`)
         .reply(200, []);
 
-      const res = await github.getDigest({ packageName }, null);
+      const res = await github.getDigest({ packageName }, null as never);
```
Comment (viceice) [inline lib/modules/datasource/github-tags/index.spec.ts:54]:
```suggestion
      const res = await github.getDigest({ packageName }, undefined);
```

File: lib/modules/datasource/github-tags/index.spec.ts [hunk 12]
```diff
@@ -86,7 +86,7 @@ describe('modules/datasource/github-tags/index', () => {
 
       const res = await github.getDigest(
         { packageName, registryUrl: githubEnterpriseApiHost },
-        null
+        null as never
```
Comment (viceice) [inline lib/modules/datasource/github-tags/index.spec.ts:89]:
```suggestion
        undefined
```

File: lib/modules/datasource/go/releases-goproxy.spec.ts [hunk 13]
```diff
@@ -76,7 +76,7 @@ describe('modules/datasource/go/releases-goproxy', () => {
 
     it('ignores everything starting from "direct" and "off" keywords', () => {
       expect(datasource.parseGoproxy(undefined)).toBeEmpty();
-      expect(datasource.parseGoproxy(null)).toBeEmpty();
+      expect(datasource.parseGoproxy(null as never)).toBeEmpty();
```
Comment (viceice) [inline lib/modules/datasource/go/releases-goproxy.spec.ts:79]:
```suggestion
      expect(datasource.parseGoproxy(undefined)).toBeEmpty();
```

**Review by viceice — COMMENTED** [review]
some suggestions

File: lib/modules/datasource/conan/index.spec.ts [hunk 14]
```diff
@@ -31,7 +31,7 @@ describe('modules/datasource/conan/index', () => {
       httpMock
         .scope(nonDefaultRegistryUrl)
         .get('/v2/conans/search?q=fakepackage')
-        .reply(200, null);
+        .reply(200, null as never);
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/datasource/conan/index.spec.ts:34]:
```suggestion
        .reply(200);
```
Will this work?

[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/modules/datasource/conan/index.spec.ts:34]: it should.

**Review by viceice — APPROVED** [review]

**Pull request merged**